Avoid unnecessary StyleColor copying in animation wrappers

anttijk opened the pull request:
#### 7baa7bfbdcd3a44864124f78e7b67bf73f868722
<pre>
Avoid unnecessary StyleColor copying in animation wrappers
<a href="https://bugs.webkit.org/show_bug.cgi?id=255278">https://bugs.webkit.org/show_bug.cgi?id=255278</a>
rdar://107875297

Reviewed by Antoine Quint.

There is some unnecessary copying of this not-quite-trivial type.

* Source/WebCore/animation/CSSPropertyAnimation.cpp:
(WebCore::PropertyWrapperStyleColor::PropertyWrapperStyleColor):
(WebCore::PropertyWrapperStyleColor::value const):

Remove templating and have separate classes for StyleColor and Color cases.

(WebCore::PropertyWrapperColor::PropertyWrapperColor):
(WebCore::PropertyWrapperColor::value const):
(WebCore::PropertyWrapperVisitedAffectedStyleColor::PropertyWrapperVisitedAffectedStyleColor):
(WebCore::PropertyWrapperVisitedAffectedColor::PropertyWrapperVisitedAffectedColor):
(WebCore::CSSPropertyAnimationWrapperMap::CSSPropertyAnimationWrapperMap):
(WebCore::PropertyWrapperColor::unresolvedValue const): Deleted.

Canonical link: <a href="https://commits.webkit.org/262822@main">https://commits.webkit.org/262822@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/601bc4209665929c42e8abf7f280f0e64c328faa

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/2705 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/2706 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/2847 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/4113 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/3075 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/2669 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/2849 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/2802 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/2381 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/2730 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/3172 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/2418 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/3880 "Built successfully") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/39/builds/666 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/2403 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/2262 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/2787 "Built successfully") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/2422 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/3634 "Passed tests") | 
| | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/2454 "Passed tests") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/2235 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/2457 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/2406 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/667 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/2447 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/2601 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

graouts commented on Source/WebCore/animation/CSSPropertyAnimation.cpp:
```diff
@@ -1695,57 +1695,89 @@ class PropertyWrapperShadow final : public AnimationPropertyWrapperBase {
+    void blend(RenderStyle& destination, const RenderStyle& from, const RenderStyle& to, const CSSPropertyBlendingContext& context) const override
+    {
+        auto result = blendFunc(value(from), value(to), context);
+        (destination.*m_setter)(result);
```
Bit sad that we can't move this. I guess it gets copied down the line?

anttijk replied:
I think we can.

anttijk committed:
Avoid unnecessary StyleColor copying in animation wrappers
https://bugs.webkit.org/show_bug.cgi?id=255278
rdar://107875297

Reviewed by Antoine Quint.

There is some unnecessary copying of this not-quite-trivial type.

* Source/WebCore/animation/CSSPropertyAnimation.cpp:
(WebCore::PropertyWrapperStyleColor::PropertyWrapperStyleColor):
(WebCore::PropertyWrapperStyleColor::value const):

Remove templating and have separate classes for StyleColor and Color cases.

(WebCore::PropertyWrapperColor::PropertyWrapperColor):
(WebCore::PropertyWrapperColor::value const):
(WebCore::PropertyWrapperVisitedAffectedStyleColor::PropertyWrapperVisitedAffectedStyleColor):
(WebCore::PropertyWrapperVisitedAffectedColor::PropertyWrapperVisitedAffectedColor):
(WebCore::CSSPropertyAnimationWrapperMap::CSSPropertyAnimationWrapperMap):
(WebCore::PropertyWrapperColor::unresolvedValue const): Deleted.

Canonical link: https://commits.webkit.org/262822@main
